sound/soc/sof/topology.c

```diff
@@ -2019,6 +2042,108 @@ static int sof_get_control_data(struct snd_soc_component *scomp,
 	return 0;
 }
 
+/*
+ * Sof mux/demux component
+ */
+static int sof_process_load_mux(struct snd_soc_component *scomp,
+				struct sof_ipc_comp_process *process,
+				struct snd_soc_tplg_private *private,
+				int *offset)
+{
+	struct sof_ipc_stream_map *sm =
+		(struct sof_ipc_stream_map *)((u8 *)process->data + (*offset));
+	u8 *pos = (u8 *)process->data + (*offset) +
+		sizeof(struct sof_ipc_stream_map);
+	struct sof_ipc_channel_map *cm_temp = NULL;
+	struct sof_ipc_channel_map *cm_temp_p;
+	u32 *cm_coeffs_temp = NULL;
+	u32 *cm_coeffs_temp_p;
+	int num_coeffs = 0;
+	size_t size;
+	int ret;
+	int i;
+
+	/* parse stream map */
+	ret = sof_parse_tokens(scomp, sm, stream_map_tokens,
+			       ARRAY_SIZE(stream_map_tokens), private->array,
+			       le32_to_cpu(private->size));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: mux stream map tokens failed %d\n",
+			private->size);
+		return ret;
+	}
+
+	/* reserve temp memory for variable amount of channel maps */
+	size = sizeof(struct sof_ipc_channel_map) * sm->num_ch_map;
+	cm_temp = kzalloc(size, GFP_KERNEL);
+	if (!cm_temp)
+		return -ENOMEM;
+
+	/* parse channel maps to temp memory cm_temp */
+	ret = sof_parse_token_sets(scomp, cm_temp, channel_map_tokens,
+				   ARRAY_SIZE(channel_map_tokens),
+				   private->array,
+				   le32_to_cpu(private->size),
+				   sm->num_ch_map,
+				   sizeof(struct sof_ipc_channel_map));
+	if (ret != 0) {
+		dev_err(scomp->dev, "error: mux channel map tokens %d\n",
+			private->size);
+		goto out;
+	}
+
+	/* calculate total amount of coeffs i.e. all bits set in all masks */
+	cm_temp_p = cm_temp;
+	for (i = 0; i < sm->num_ch_map; i++, cm_temp_p++)
+		num_coeffs += __sw_hweight32(cm_temp_p->ch_mask);
```

Comment:
@juimonen this part is confusing me a bit. for each ch_map in num_ch_map, there are a set of coefficients right? So is the expectation that the coefficients are in order for each ch_map ?

Reply:
well so far there's no "id" or anything for coeffs, as they are as var array. So yes here it is assumed that coeffs are in order, so belonging to the ch_maps (in order). So with my modified parsing, I can parse all at the same time, or I could use inner loop. It doesn't change the problematics of copying them eventually to correct place in ipc struct.

Reply:
@juimonen this makes me a bit uncomfortable. What we're talking about is a set of ch_map tokens, each a with variable number of coefficients. Does the coefficients have to be a token array or can they be data with matching ch_map id?

Reply:
So I'm not sure we can mix data and tokens here, I think I tried it sometimes and I mean our parser will go berserk... So I think it is either token arrays or data (or data is in the end and you have to come up with a position to jump there). So as I understand the upper struct has the amount of coeffs, and that variable amount of coeffs is following the ch_map. I assume that first coeffs belong to first ch_map, second to second and so on... Anyway if we are getting rid of var arrays this has to change

Reply:
@juimonen token arrays are fine but we need a way to match the coefficients with the ch_map_id. I dont think the assumption that the topology will always be perfect is valid. So we parse the ch_map_tokens first and then the ch_map_coeff_tokens for each ch_map id. This will complicate the parsing function a bit but guess what the ch_map_id field in each coeff will give us the offset to where to copy them. makes sense?

Reply:
hmm I'm not sure how to do that..? doesn't this also assume that the id of the coeff is somewhere close to it (just before it) in arrays? So not sure this will make it more robust than now... or am I missing something?

Reply:
what I meant was you'll need to extend the coeff tokens to include also the ch_map_id apart from the actual coeff itself